cmd/influx_inspect/verify_seriesfile/verify_seriesfile.go

```diff
@@ -0,0 +1,298 @@
+		Stdout: os.Stdout,
+	}
+}
+
+// Run executes the command.
+func (cmd *Command) Run(args ...string) error {
+	fs := flag.NewFlagSet("verify-seriesfile", flag.ExitOnError)
+	fs.SetOutput(cmd.Stdout)
+
+	dataDir := fs.String("dir", os.Getenv("HOME")+"/.influxdb", "data directory.")
+	dbPath := fs.String("db", "", "database path. overrides data directory.")
+	filePath := fs.String("file", "", "series file path. overrides db and data directory.")
+
+	if err := fs.Parse(args); err != nil {
+		return err
+	}
+
+	return cmd.run(*dataDir, *dbPath, *filePath)
+}
+
+// run performs all of the validations, preferring a filePath, then a dbPath,
+// and then a dataDir.
+func (cmd *Command) run(dataDir, dbPath, filePath string) error {
+	if filePath != "" {
+		_, err := cmd.runFile(filePath)
+		return err
+	}
+	if dbPath != "" {
+		_, err := cmd.runFile(filepath.Join(dbPath, "_series"))
+		return err
+	}
+
+	dbs, err := ioutil.ReadDir(filepath.Join(dataDir, "data"))
+	if err != nil {
+		return err
+	}
+
+	for _, db := range dbs {
+		if !db.IsDir() {
+			continue
+		}
+		_, err := cmd.runFile(filepath.Join(dataDir, "data", db.Name(), "_series"))
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// runFile performs validations on a series file. The error is only returned if
+// there was some fatal problem with operating, not if there was a problem with
+// the series file.
+func (cmd *Command) runFile(filePath string) (valid bool, err error) {
+	logger := logger.New(cmd.Stderr).With(zap.String("path", filePath))
+	logger.Info("starting validation")
+
+	defer func() {
+		if rec := recover(); rec != nil {
+			logger.Error("panic validating file",
+				zap.String("recovered", fmt.Sprint(rec)))
+			valid = false
+		}
+	}()
+
+	partitionInfos, err := ioutil.ReadDir(filePath)
+	if os.IsNotExist(err) {
+		logger.Error("series file does not exist")
+		return false, nil
+	}
+	if err != nil {
+		return false, err
+	}
+
+	// check every partition
+	for _, partitionInfo := range partitionInfos {
+		pLogger := logger.With(zap.String("partition", partitionInfo.Name()))
+		pLogger.Info("validating partition")
+		partitionPath := filepath.Join(filePath, partitionInfo.Name())
+
+		segmentInfos, err := ioutil.ReadDir(partitionPath)
+		if err != nil {
+			return false, err
+		}
+		segments := make([]*tsdb.SeriesSegment, 0, len(segmentInfos))
+		ids := make(map[uint64]idData)
+
+		// check every segment
+		for _, segmentInfo := range segmentInfos {
+			segmentPath := filepath.Join(partitionPath, segmentInfo.Name())
+
+			segmentID, err := tsdb.ParseSeriesSegmentFilename(segmentInfo.Name())
+			if err != nil {
+				continue
+			}
+
+			sLogger := pLogger.With(zap.String("segment", segmentInfo.Name()))
+			sLogger.Info("validating segment")
+
+			segment := tsdb.NewSeriesSegment(segmentID, segmentPath)
+			if err := segment.Open(); err != nil {
+				sLogger.Error("opening segment", zap.Error(err))
+				return false, nil
+			}
+			defer segment.Close()
```

Comment:
This `Close()` won't be deferred until the very end of `runFile()`. That probably isn't a huge issue unless there are a bunch of segments and limited file descriptors. Maybe refactor this block to a `runSegment()` function so `defer` runs after ever segment?

Reply:
I actually intend for them not to be closed until the end since we use them for checking the index later. Defer in a for loop is definitely a red flag, so I'll leave a comment to avoid false positives.

Reply:
Oh nope, I'm wrong. That's an artifact from when it was just checking a single partition in a function. Thanks.